crates/iceberg/src/expr/visitors/manifest_evaluator.rs

```diff
@@ -0,0 +1,393 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::expr::visitors::bound_predicate_visitor::{visit, BoundPredicateVisitor};
+use crate::expr::visitors::inclusive_projection::InclusiveProjection;
+use crate::expr::{Bind, BoundPredicate, BoundReference};
+use crate::spec::{Datum, FieldSummary, ManifestFile, PartitionSpecRef, Schema, SchemaRef};
+use fnv::FnvHashSet;
+use std::sync::Arc;
+
+/// Evaluates [`ManifestFile`]s to see if their partition summary matches a provided
+/// [`BoundPredicate`]. Used by [`TableScan`] to filter down the list of [`ManifestFile`]s
+/// in which data might be found that matches the TableScan's filter.
+pub(crate) struct ManifestEvaluator {
+    partition_schema: SchemaRef,
+    partition_filter: BoundPredicate,
+    case_sensitive: bool,
+}
+
+impl ManifestEvaluator {
+    pub(crate) fn new(
+        partition_spec: PartitionSpecRef,
+        table_schema: SchemaRef,
+        partition_filter: BoundPredicate,
+        case_sensitive: bool,
+    ) -> crate::Result<Self> {
+        let partition_type = partition_spec.partition_type(&table_schema)?;
+
+        // this is needed as SchemaBuilder.with_fields expects an iterator over
+        // Arc<NestedField> rather than &Arc<NestedField>
+        let cloned_partition_fields: Vec<_> =
+            partition_type.fields().iter().map(Arc::clone).collect();
+
+        let partition_schema = Schema::builder()
+            .with_fields(cloned_partition_fields)
+            .build()?;
+
+        let partition_schema_ref = Arc::new(partition_schema);
+
+        let mut inclusive_projection = InclusiveProjection::new(partition_spec.clone());
+        let unbound_partition_filter = inclusive_projection.project(&partition_filter)?;
+
+        let partition_filter =
+            unbound_partition_filter.bind(partition_schema_ref.clone(), case_sensitive)?;
```

Comment:
in the python impl `rewrite_not` is also applied: https://github.com/apache/iceberg-python/blob/main/pyiceberg/expressions/visitors.py#L555
...however, I'm not sure if at this point our ExprTree could even contain any NOT nodes?

Reply:
Similarly with InclusiveProjection, since we apply rewrite-not when a Predicate is passed to the TableScan as it gets built, we've already guaranteed that there are no Not nodes by the time we get here

Reply:
This may not be a blocker, but I think adding a rewrite not phase would be safer. At least we will throw error rather than panic.

Reply:
Ok, I suppose it's not exactly a huge overhead :-)